### PR TITLE
Fix `Camera3D` projection issues with `PROJECTION_FRUSTUM`, viewport 2d override and `KEEP_WIDTH`

### LINKED ISSUE #61174: Camera3D project_* methods not accounting for frustum offset
### Godot version

4.0.dev

### System information

Arch Linux 5.16.14

### Issue description

When using a `Camera3D`'s `project_ray_origin`, `project_ray_normal`, `project_position` and `unproject_position` to translate between world and screen coordinates, if the camera is set to frustum projection, the projected position does not match the screen position.
In this example I set up two viewports, one with a frustum camera and an one with an overview perspective camera for debugging. I had a simple script converting positions from `InputEventMouseButton`s into ray origin and normal, then using a `RayCast3D` to intersect with the scene (the three prisms have static bodies)

The expected result (and the one I got with perspective projection, orthographic projection and frustum projection with no offset) is that the green cylinder (that indicates the world position gotten from the conversion) should point to the box in the back, but instead it points at the wall.

Looking into the source, the error is obvious. None of the methods handles frustum projections, there's just a 
```c++
if (mode == PROJECTION_PERSPECTIVE) {
    // stuff for perspective
} else {
    // stuff for orthographic (and implicitly frustum)
}
```
or
```c++
if (mode == PROJECTION_ORTHOGRAPHIC) {
    // stuff for orthographic
} else {
    // stuff for perspective (and implicitly frustum)
}
```

### Steps to reproduce

Set up a 3D scene with a frustum camera with a non-zero offset. Set up a script that uses a world-space position generated from to screen positions using one of the aforementioned methods. I recommend moving a 3D object to the position gotten from a click, as it's the easiest way to see the difference.

### Minimal reproduction project

Sorry for the messy layout. It's the same project as the screenshot.
Change the settings of the `Camera3D` in `Node3D.tscn` to test that indeed it does work on

[debug_program.zip](https://github.com/godotengine/godot/files/8721829/debug_program.zip)

### LINKED ISSUE #77906: unproject_position and is_position_in_frustum returns wrong results if window is resized to a different aspect ratio
### Godot version

4.0.3

### System information

Windows 10 - Godot v4.0.3.stable - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1070 (Game Ready Version 528.49) - Intel Core i7-7700K CPU @ 4.20GHz (8 Threads)

### Issue description

Problem:
I get wrong results for unproject_position and is_position_in_frustum when the aspect ratio is not the same as set in the project settings' viewport size.

As you can see in my settings, I have 1:1 aspect ratio in project settings' viewport size, in order to support both landscape and portrait screens. It's a recommended setting in one of the green tip boxes at https://docs.godotengine.org/en/stable/tutorials/rendering/multiple_resolutions.html. Which also seems to be working the best for keeping relative sizes of GUI elements for different aspect ratios. But I'm using 1.77 aspect ratio in project settings' viewport override size, so the game's initial window can fit my own screen during development. And this is how I encountered the problem.

Although the 3D object is visible on screen, is_position_in_frustum returns false if the object is a bit far on left or right side of the viewport. And unproject_position for global_transform.origin can return even negative values on left parts of the viewport, and it doesn't match the correct coordinates on the right side either. It behaves as if the window is narrower on x axis.

If I change camera's "Keep Aspect" value from keep height to keep width, then the functions start working correctly for x axis, but they behave as if the viewport is higher, like, is_position_in_frustum keeps returning true till some point even when the object is vertically outside of the viewport, like very below or above and not visible.

The functions return correct results only if the current aspect ratio of the window is the same with the viewport size in the project settings.

What do I expect?
I expect the functions to return correct results, regardless of the aspect ratio being same or not with the viewport size set in the project settings.

### Steps to reproduce

My settings:

PROJECT
viewport width 1920
viewport height 1920
window width override 640
window height override 400
resizable true
mode viewport
aspect expand
scale 1
orientation sensor

SUBVIEWPORTCONTAINER
stretch true
stretch shrink 1

SUBVIEWPORT
size 1920 1920
size 2d override 1920 1920
size 2d override stretch true

CAMERA
keep aspect keep height
projection perspective

My node structure:

Control
- SubViewportContainer
- - SubViewport
- - - Camera3D
- - - MeshInstance3D
- Control (for 2D interface elements...)

My logic:

The mesh instance in subject never moves, and my camera has RTS like angle. I use _physics_process(delta), and I simply make the camera move left, right, forward and backward with W, A, S, D and print(camera.is_position_in_frustum(subject.global_transform.origin), camera.unproject_position(subject.global_transform.origin), subviewport.size, subviewport.size_2d_override, get_viewport().size, get_viewport().get_visible_rect().size, get_viewport_rect().size, get_global_rect().size). Then I just compare the values, but they don't match with anything.

Unless my window's aspect ratio is the same with the viewport size in the project settings, I keep getting wrong results on x axis. And if I change the camera's Keep Aspect value from keep height to keep width, then I keep getting wrong results on y axis.

### Minimal reproduction project

[terminal.zip](https://github.com/godotengine/godot/files/11663791/terminal.zip)

### LINKED ISSUE #90074: Camera3D.unproject_offset does not work with FRUSTRUM_PROJECTION projection type.
### Tested versions

- Reproducible in 4.x

### System information

Windows 11 - Godot 4.1-stable

### Issue description

Cameras with frustrum projection don't work properly with unproject_offset. I've tried looking around to see if anyone else is having this issue. Unfortunately for me frustrum projection is a kind of obscure feature and I'm using it in doubly obscure ways. The most helpful document I ran into was [Deprecate Camera3D's PROJECTION_FRUSTUM projection mode](https://github.com/godotengine/godot-proposals/issues/8536), which explained some of the known jank. Ideally, I would be using custom projection matrices. For now, I'd be satisfied with just a tip to get the screen position under frustrum projection, but it would be nice if the function worked with frustrum projection out of the box.

### Steps to reproduce

Download the MRP. Launch the test scene and rotate the camera with the arrow keys. If the camera's project is changed to perspective or orthographic, the sprite will be placed over the cube (as long as the cube is on screen, I didn't bother adding checks otherwise for the mrp). In frustrum project though, the sprite gets offset. It appears that unproject_offset is basing it's calculations off of the fov variable, which would line up with the claim from the deprecation request: "None of the code paths in Camera3D actually explicitly handles PROJECTION_FRUSTUM."

### Minimal reproduction project (MRP)

[CameraProjectionTest.zip](https://github.com/godotengine/godot/files/14816963/CameraProjectionTest.zip)

### PR DESCRIPTION
**Pull request by Flarkk**
Resolve  #77906
Resolve #90074
Resolve #61174

Partially address the limitations mentioned in https://github.com/godotengine/godot-proposals/issues/8536

Depends on #99961 (more unit tests)

Please test it extensively.
I wrote quite a lot of unit tests but it's error prone mathy stuff so the more testers the better.

### CODE REVIEW

**Comment by Limeth** [comment]
Hello! I just ran into the issue this PR is trying to fix, and was delighted to see your efforts.
I've tried running my project with these changes, and it mostly seems to work fine, except when `Camera3D::keep_aspect` is set to `Camera3D.KeepAspect.KEEP_WIDTH` rather than the default `KEEP_HEIGHT`, then the math seems wrong. Looking at the changes, it is not yet clear to me why.

**Comment by Flarkk** [comment]
@Limeth which function is affected ?

**Comment by Limeth** [comment]
> @Limeth which function is affected ?

Forgive my unfamiliarity with Godot, but I am seeing the issue on `_mouse_enter()` and `_mouse_exit()` of an `Area3D` node.
I suppose it must be using `project_local_ray_normal` to ray-trace the scene and trigger the event listeners above.

**Comment by Limeth** [comment]
A demo with a demonstration.

Correct behavior with `KEEP_HEIGHT`:
https://github.com/user-attachments/assets/dd767c20-3a08-4777-950e-a6b91838923b

Incorrect behavior with `KEEP_WIDTH`:
https://github.com/user-attachments/assets/00f33176-a3c2-446c-97f9-d4beaa860609

**Comment by Flarkk** [comment]
Just fixed a couple of issues with the aforementioned functions not taking into account `KEEP_WIDTH` cameras.
More importantly I reworked the test to facilitate programmatical testing of various combinations of `Camera3D.KeepAspect`, `Viewport.size`, `Viewport.size_2d_override`.
This should help catching more issues.
@Limeth it would be great if you could test again.

**Comment by Limeth** [comment]
I tested my project with the changes, and unfortunately it seems that neither of the keep_aspect options are working for me now.
Since my project is pretty small so far, I thought it'd be best for me to share it so you can investigate what the issue might be yourself.

[project.zip](https://github.com/user-attachments/files/18101414/project.zip)

You can do the same test as in the videos above just by running the project and holding the right mouse button.
It seems to work, as long as the window isn't resized. Resizing the window changes the keep_aspect option, depending on whether the aspect ratio increased or decreased compared to the original aspect ratio, to always keep the scene within the window.
Resizing also seems to break the ray tracing math.

Let me know if there's anything more I could do to help out.

**Comment by Flarkk** [comment]
Bad news... But thanks for taking time. It helps. Will take a look again

**Comment by Flarkk** [comment]
Here we are ! I made a small mistake both in the code and in the tests.
I could test it extensively with both your project and the test scene from #61174, with various combinations of `KEEP_HEIGHT`/`KEEP_WIDTH`, viewport size and viewport size override. 
I think I finally got it right.

All this made me realize that `SubViewport.size_2d_override_stretch` doesn't work [as documented](https://docs.godotengine.org/en/stable/classes/class_subviewport.html#class-subviewport-property-size-2d-override-stretch). It's rather a global flag that toggles on / off the whole 2d size overriding logic - not only the stretch aspect of it.

In other words, when `size_2d_override_stretch` is `false`, `size_2d_override` is ignored everywhere.
Note that the base class' `Viewport.size_2d_override_stretch()` always return true and `Window` inherits this behavior. In For those 2 classes `size_2d_override` will be effective right away.

We might want to clarify the documentation at some point.

**Comment by Limeth** [comment]
I can confirm that it works properly for me now. Thank you!

**Comment by Flarkk** [comment]
@lyuma would you have time to take a look at this ? Cc @fire

**Comment by l0g0sys** [comment]
@Flarkk Is there programmatic workaround for this? I need to use unproject_position() on camera with Frustrum projection and KEEP_WIDTH mode.

**Comment by Flarkk** [comment]
I think it's doable to reimplement `unproject_position()` in GDScript, and call from there the fixed version of `_get_camera_projection()` in GDScript too :

```c++
Point2 Camera3D::unproject_position(const Vector3 &p_pos) const {
	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector2(), "Camera is not inside scene.");

	Size2 viewport_size = get_viewport()->get_visible_rect().size;

	Projection cm = _get_camera_projection(_near); // This must use the fixed version

	Plane p(get_camera_transform().xform_inv(p_pos), 1.0);

	p = cm.xform4(p);

	// Prevent divide by zero.
	// TODO: Investigate, this was causing NaNs.
	ERR_FAIL_COND_V(p.d == 0, Point2());

	p.normal /= p.d;

	Point2 res;
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x;
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y;

	return res;
}

```

```c++
Projection Camera3D::_get_camera_projection(real_t p_near) const {
	Size2 viewport_size = get_viewport()->get_camera_rect_size();
	Projection cm;

	switch (mode) {
		case PROJECTION_PERSPECTIVE: {
			cm.set_perspective(fov, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH);
		} break;
		case PROJECTION_ORTHOGONAL: {
			cm.set_orthogonal(size, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH);
		} break;
		case PROJECTION_FRUSTUM: {
			cm.set_frustum(size, viewport_size.aspect(), frustum_offset, p_near, _far, keep_aspect == KEEP_WIDTH);
		} break;
	}

	return cm;
}
```

**Comment by l0g0sys** [comment]
Yep, looks like lot of fun.
Already re-implemented original unproject_position() along with unexposed C++ methods like xform_inv() and xform4(), but I had to make mistake somewhere as it doesn't return same coords as original one.

**Comment by Flarkk** [comment]
If you can share your code I may be able to help.

**Comment by l0g0sys** [comment]
So, this is original Camera3D.UnprojectPosition():
```
	private static Vector2 UnprojectPosition(Camera3D camera, Vector3 p_pos)
	{
		Vector2 viewport_size = camera.GetViewport().GetVisibleRect().Size;

		Projection cm = camera.GetCameraProjection();

		Plane p = new Plane(XformInv(camera.Transform, p_pos), 1.0f);

		p = Xform4(cm, p);
		p.Normal /= p.D;

		return new Vector2(
			(p.Normal.X * 0.5f + 0.5f) * viewport_size.X,
			(-p.Normal.Y * 0.5f + 0.5f) * viewport_size.Y
		);
	}
```
Internal C++ Vector3 Transform3D::xform_inv(const Vector3 &p_vector):
```
	private static Vector3 XformInv(Transform3D transform, Vector3 p_vector)
	{
		Vector3 v = p_vector - transform.Origin;

		Basis basis = transform.Basis;

		return new Vector3(
			(basis.Row0[0] * v.X) + (basis.Row1[0] * v.Y) + (basis.Row2[0] * v.Z),
			(basis.Row0[1] * v.X) + (basis.Row1[1] * v.Y) + (basis.Row2[1] * v.Z),
			(basis.Row0[2] * v.X) + (basis.Row1[2] * v.Y) + (basis.Row2[2] * v.Z)
		);
	}

```
And internal C++ Plane Projection::xform4(const Plane &p_vec4):
```
	private static Plane Xform4(Projection projection, Plane p_vec4)
	{
		Vector3 normal = new Vector3(
			projection[0][0] * p_vec4.Normal.X + projection[1][0] * p_vec4.Normal.Y + projection[2][0] * p_vec4.Normal.Z + projection[3][0] * p_vec4.D,
			projection[0][1] * p_vec4.Normal.X + projection[1][1] * p_vec4.Normal.Y + projection[2][1] * p_vec4.Normal.Z + projection[3][1] * p_vec4.D,
			projection[0][2] * p_vec4.Normal.X + projection[1][2] * p_vec4.Normal.Y + projection[2][2] * p_vec4.Normal.Z + projection[3][2] * p_vec4.D
		);

		float d = projection[0][3] * p_vec4.Normal.X + projection[1][3] * p_vec4.Normal.Y + projection[2][3] * p_vec4.Normal.Z + projection[3][3] * p_vec4.D;

		return new Plane(normal, d);
	}
```

**Comment by Flarkk** [comment]
> `Projection cm = camera.GetCameraProjection();`

You need to reimplement `GetCameraProjection()` too. This is what this PR ultimately fixes.

**Comment by l0g0sys** [comment]
Yeah, i will. But first I have to reimplement old code so I can verify it against real UnprojectPosition method. After I get it working 1:1, then I can apply fix.

**Comment by l0g0sys** [comment]
Oh, found the error.
 On line `Plane p = new Plane(XformInv(camera.Transform, p_pos), 1.0f);`
value of `camera.Transform` is different than value of C++ `get_camera_transform()` method.

**Comment by Flarkk** [comment]
If you can share your final workaround code in this thread it may be found useful by other users until this PR gets merged.

**Comment by l0g0sys** [comment]
Updated those reimplementations above to include also workaround. Didn't want to spam extra set of code.

**Comment by Flarkk** [comment]
Rebased using `Math` namespace for `SQRT3` math constant.

**Comment by Flarkk** [comment]
Unit tests this PR adds that cover `FRUSTUM` camera mode now fail since this rebase.

I couldn't spend time on it yet. There are 3 possible causes:

1. Precision issues in the tests introduced by `Math::SQRT3` (which is a `constexpr double` unlike the previous local `#define SQRT3 1.732050f`)

2. A mistake a could have made doing the rebase that could have changed the tests' logic

3. A regression affecting projections in `FRUSTUM` camera mode that were introduced by another PR in the meantime

I will switch this PR to draft until it's resolved.

**Comment by Flarkk** [comment]
Now tests pass.
It was case no. 3 with `Projection::get_viewport_half_extents()` having a different behavior with non-symmetrical viewports since #100209.

There is now another unrelated compilation error with `Vector2` multiplication I didn't experienced locally.
I should be able to fix it shortly.

**Comment by Flarkk** [comment]
Here we are.
Reviews welcome !

**Comment by Flarkk** [comment]
Thanks for reporting. If you can somehow provide a MRP, that would help. Will wrap my mind around what you describe anyway.

**Comment by Flarkk** [comment]
In your video it looks like that it acts as if the viewport had twice smaller extent from the top-left corner. For instance when the label should be at the center, it's actually positioned on the bottom-right corner. Is this what you observed ?

**Comment by Flarkk** [comment]
@benbot have you been able to narrow it down or to make a MRP ?

**Comment by Flarkk** [comment]
I tried to reproduce what you describe but haven't been very successful so far.
I got somewhat similar behavior with the following scene setup, when `stretch` is **not** enabled on the `SubViewportContainer`. Ticking `stretch` solves the problem. Have your scene a similar setup ?
image

**Comment by elvisish** [comment]
Is it possible to get a fix for 3.6 or 3.7? I'm having to do the calculations with magic numbers:
```
(cam.project_ray_origin(viewport.size * 0.5) + cam.project_ray_normal(viewport.size * 0.5 + Vector2(0,cam.get_frustum_offset().y * -20000)) * ray_length)
```
but unfortunately this breaks if I change the viewport size or size of the frustum camera.

**Comment by Flarkk** [comment]
> Is it possible to get a fix for 3.6 or 3.7?

I don't see a reason this PR cannot be cherry-picked in 3.x.

If it turns out to require a custom implementation I can contribute.

**Comment by elvisish** [comment]
> > Is it possible to get a fix for 3.6 or 3.7?
> 
> I don't see a reason this PR cannot be cherry-picked in 3.x.
> 
> If it turns out to require a custom implementation I can contribute.

Thankyou, I think @lawnjelly is the sole maintainer for 3.x, I'm trying to finish up a game I've been working on for 3 years now but after reworking the viewport setup it's broken the hack I was using for this.

Can it be implemented in gdscript currently so I can get frustum offset ray projection as close as possible until it might potentially be backported? I'm thinking about trying some kind of weird hacky rotation gimbal temporarily to get a close enough position, but I'd much rather just try a new projection function that actually takes the offset into account.

**Comment by Flarkk** [comment]
> Can it be implemented in gdscript currently so I can get frustum offset ray projection as close as possible until it might potentially be backported?

Yes it can be worked around in GDScript. See my [above post](https://github.com/godotengine/godot/pull/100121#issuecomment-2763233784).

**Comment by elvisish** [comment]
> > Can it be implemented in gdscript currently so I can get frustum offset ray projection as close as possible until it might potentially be backported?
> 
> Yes it can be worked around in GDScript. See my [above post](https://github.com/godotengine/godot/pull/100121#issuecomment-2763233784).

I saw that, but it kinda looks more like cpp to me? Also, I'm using project_ray_origin and project_ray_normal to test a ray forwards from the camera in the direction of the frustum offset, would those functions be much different reimplemented?

**Comment by elvisish** [comment]
> > Can it be implemented in gdscript currently so I can get frustum offset ray projection as close as possible until it might potentially be backported?
> 
> Yes it can be worked around in GDScript. See my [above post](https://github.com/godotengine/godot/pull/100121#issuecomment-2763233784).

I've tried to port this to GDScript, but it's not quite working:
```swift
func unproject_position(p_pos):
	print(!is_inside_tree(), Vector2(), "Camera is not inside scene.")

	var viewport_size : Vector2 = get_viewport().get_visible_rect().size

	var cm = _get_camera_projection(_near) ## This must use the fixed version

	p(get_camera_transform().xform_inv(p_pos), 1.0)

	p = cm.xform4(p);

	## Prevent divide by zero.
	## TODO: Investigate, this was causing NaNs.
	ERR_FAIL_COND_V(p.d == 0, Point2());

	p.normal /= p.d;

	var res
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y

	return res

func _get_camera_projection(real_t, p_near):
	var viewport_size : Rect2 = get_viewport().get_camera_rect_size()
	var cm

	match (mode):
		PROJECTION_PERSPECTIVE: 
			cm.set_perspective(fov, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_ORTHOGONAL: 
			cm.set_orthogonal(size, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_FRUSTUM: 
			cm.set_frustum(size, viewport_size.aspect(), frustum_offset, p_near, _far, keep_aspect == KEEP_WIDTH)

	return cm
```
I'm kinda desperate at this point, but I still need `project_ray_origin` and `project_ray_normal` to be ported and I'm not sure where to start with those.

**Comment by Flarkk** [comment]
> I still need project_ray_origin and project_ray_normal to be ported and I'm not sure where to start with those.

As far as I remember `project_ray_origin` is not affected by this PR because it always returns the camera origin in Frustum and Perspective modes. You should be able to use the vanilla version here.

For `project_ray_normal` you need to rewrite the following in GDScript : 

```c++
Vector3 Camera3D::project_ray_normal(const Point2 &p_pos) const {
	Vector3 ray = project_local_ray_normal(p_pos);
	return get_camera_transform().basis.xform(ray).normalized();
}

Vector3 Camera3D::project_local_ray_normal(const Point2 &p_pos) const {
	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector3(), "Camera is not inside scene.");

	Size2 viewport_size = get_viewport()->get_camera_rect_size();
	Vector2 cpos = get_viewport()->get_camera_coords(p_pos);
	Vector3 ray;

	if (mode == PROJECTION_ORTHOGONAL) {
		ray = Vector3(0, 0, -1);
	} else {
		Projection cm = _get_camera_projection(_near); // This must be the fixed version 
		Vector2 top_right = cm.get_viewport_half_extents();
		Vector2 offset = Vector2();
		if (mode == PROJECTION_FRUSTUM) {
			offset = frustum_offset;
		}
		ray = Vector3(((cpos.x / viewport_size.width) * 2.0 - 1.0) * top_right.x + offset.x, ((1.0 - (cpos.y / viewport_size.height)) * 2.0 - 1.0) * top_right.y + offset.y, -_near).normalized();
	}

	return ray;
}
```

**Comment by elvisish** [comment]
> > I still need project_ray_origin and project_ray_normal to be ported and I'm not sure where to start with those.
> 
> As far as I remember `project_ray_origin` is not affected by this PR because it always returns the camera origin in Frustum and Perspective modes. You should be able to use the vanilla version here.
> 
> For `project_ray_normal` you need to rewrite the following in GDScript :
> 
> ```c++
> Vector3 Camera3D::project_ray_normal(const Point2 &p_pos) const {
> 	Vector3 ray = project_local_ray_normal(p_pos);
> 	return get_camera_transform().basis.xform(ray).normalized();
> }
> 
> Vector3 Camera3D::project_local_ray_normal(const Point2 &p_pos) const {
> 	ERR_FAIL_COND_V_MSG(!is_inside_tree(), Vector3(), "Camera is not inside scene.");
> 
> 	Size2 viewport_size = get_viewport()->get_camera_rect_size();
> 	Vector2 cpos = get_viewport()->get_camera_coords(p_pos);
> 	Vector3 ray;
> 
> 	if (mode == PROJECTION_ORTHOGONAL) {
> 		ray = Vector3(0, 0, -1);
> 	} else {
> 		Projection cm = _get_camera_projection(_near); // This must be the fixed version 
> 		Vector2 top_right = cm.get_viewport_half_extents();
> 		Vector2 offset = Vector2();
> 		if (mode == PROJECTION_FRUSTUM) {
> 			offset = frustum_offset;
> 		}
> 		ray = Vector3(((cpos.x / viewport_size.width) * 2.0 - 1.0) * top_right.x + offset.x, ((1.0 - (cpos.y / viewport_size.height)) * 2.0 - 1.0) * top_right.y + offset.y, -_near).normalized();
> 	}
> 
> 	return ray;
> }
> ```

I've been struggling with this, here's what I have so far:
```swift
extends Camera

func project_ray_normal(p_pos):
	var ray : Vector3 = project_local_ray_normal(p_pos)
	return get_camera_transform().basis.xform(ray).normalized()


func project_local_ray_normal(p_pos):
	print(!is_inside_tree(), Vector3(), "Camera is not inside scene.")
	var viewport_size : Rect2 = get_viewport().get_visible_rect()
	var cpos : Vector2 = get_viewport().get_camera_coords(p_pos)
	var ray : Vector3

	if (projection == PROJECTION_ORTHOGONAL):
		ray = Vector3(0, 0, -1)
	else:
		var cm = _get_camera_projection(near) ## This must be the fixed version 
		var top_right : Vector2 = cm.get_viewport_half_extents()
		var offset : Vector2
		if (projection == PROJECTION_FRUSTUM):
			offset = frustum_offset
		ray = Vector3(((cpos.x / viewport_size.size.x) * 2.0 - 1.0) * top_right.x + offset.x, ((1.0 - (cpos.y / viewport_size.size.y)) * 2.0 - 1.0) * top_right.y + offset.y, -near).normalized()
	return ray

func unproject_position(p_pos):
	print(!is_inside_tree(), Vector2(), "Camera is not inside scene.")

	var viewport_size : Vector2 = get_viewport().get_visible_rect().size

	var cm = _get_camera_projection(near) ## This must use the fixed version

	var p(get_camera_transform().xform_inv(p_pos), 1.0)

	p = cm.xform4(p);

	## Prevent divide by zero.
	## TODO: Investigate, this was causing NaNs.
	print(p.d == 0, Point2())

	p.normal /= p.d;

	var res
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y

	return res

func _get_camera_projection(p_near):
	var viewport_size : Rect2 = get_viewport().get_camera_rect_size()
	var cm

	match (projection):
		PROJECTION_PERSPECTIVE: 
			cm.set_perspective(fov, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_ORTHOGONAL: 
			cm.set_orthogonal(size, viewport_size.aspect(), p_near, _far, keep_aspect == KEEP_WIDTH)
		PROJECTION_FRUSTUM: 
			cm.set_frustum(size, viewport_size.aspect(), frustum_offset, p_near, _far, keep_aspect == KEEP_WIDTH)

	return cm
```
I'm not sure what `p(get_camera_transform().xform_inv(p_pos), 1.0)` is mean to be and I can't find a substitute for `viewport_size.aspect()`. For my purposes, I'm using a subviewport (viewportcontainer in 3.6) so maybe I could adjust that to match, but I'd still appreciate if anyone could convert this into actual working GDScript as I'm not sure if this is going to work very well, even if I can figure out how to finish it.

**Comment by Flarkk** [comment]
> I'm not sure what p(get_camera_transform().xform_inv(p_pos), 1.0) is mean to be and I can't find a substitute for viewport_size.aspect()

`p` (a `Plane` in c++) should be a `Vector4` in GDScript. Replace `p.d` by `p.w` and `p.normal` by `Vector3(p.x, p.y, p.z)`.

`viewport_size` can be converted into a `Vector2` such that you can call `Vector2.aspect()` is GDScript.

**Comment by elvisish** [comment]
> Vector3(p.x, p.y, p.z)

I'm converting to Godot 3.6 which doesn't have Vector4?

`var p(get_camera_transform().xform_inv(p_pos), 1.0)`

What should this be replaced with? And am I replacing:
```swift
p.normal /= p.d;

	var res
	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x
	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y
```
With:
```swift
Vector3(p.x, p.y, p.z) /= Vector3(p.x, p.y, p.z);

	var res
	res.x = (Vector3(p.x, p.y, p.z).x * 0.5 + 0.5) * viewport_size.x
	res.y = (-Vector3(p.x, p.y, p.z)y * 0.5 + 0.5) * viewport_size.y
```
I'm not entirely sure what those should be?

**Comment by l0g0sys** [comment]
> > Vector3(p.x, p.y, p.z)
> 
> I'm converting to Godot 3.6 which doesn't have Vector4?
> 
> `var p(get_camera_transform().xform_inv(p_pos), 1.0)`
> 
> What should this be replaced with? And am I replacing:
> 
> ```swift
> p.normal /= p.d;
> 
> 	var res
> 	res.x = (p.normal.x * 0.5 + 0.5) * viewport_size.x
> 	res.y = (-p.normal.y * 0.5 + 0.5) * viewport_size.y
> ```
> 
> With:
> 
> ```swift
> Vector3(p.x, p.y, p.z) /= Vector3(p.x, p.y, p.z);
> 
> 	var res
> 	res.x = (Vector3(p.x, p.y, p.z).x * 0.5 + 0.5) * viewport_size.x
> 	res.y = (-Vector3(p.x, p.y, p.z)y * 0.5 + 0.5) * viewport_size.y
> ```
> 
> I'm not entirely sure what those should be?

Godot 3.4 already has Plane variant. Use that.
Try to rewrite my C# Godot script code to GDScript, not C++ source code as I use native Godot type exposed for scripting.

Your code in func unproject_position:
```
var p(get_camera_transform().xform_inv(p_pos), 1.0)
```
doesn't make much sense.
Think it should be
```
var p = Plane(get_camera_transform().xform_inv(p_pos), 1.0);
```
Tho I don't know much about GDScript and its OOP syntax.

**Comment by l0g0sys** [comment]
This is what AI spewed out when i let it to rewrite my C# script to GDScript.
So you can start there.
You should be able to call statically Workaround.unproject_position(camera3d_node, p_pos_vector3) method.
[Workaround-gd.txt](https://github.com/user-attachments/files/22290190/Workaround-gd.txt)

**Comment by elvisish** [comment]
> This is what AI spewed out when i let it to rewrite my C# script to GDScript. So you can start there. You should be able to call statically Workaround.unproject_position(camera3d_node, p_pos_vector3) method. [Workaround-gd.txt](https://github.com/user-attachments/files/22290190/Workaround-gd.txt)

I keep having trouble figuring out what `Projection` type should be in 3.6:
image

**Comment by elvisish** [comment]
> This is what AI spewed out when i let it to rewrite my C# script to GDScript. So you can start there. You should be able to call statically Workaround.unproject_position(camera3d_node, p_pos_vector3) method. [Workaround-gd.txt](https://github.com/user-attachments/files/22290190/Workaround-gd.txt)

I've replaced `Projection` with `Plane` in 3.6:
image

But I'm having issues with that:
image
Should I convert all of the `[0]`,`[1]` and `[3]` to `[d]`,`[x]`,`[y]`?

**Comment by l0g0sys** [comment]
No. You cannot replace any type. All types are correct in that GD script. _get_camera_projection method must return Projection.

**Comment by elvisish** [comment]
> No. You cannot replace any type. All types are correct in that GD script. _get_camera_projection method must return Projection.

I don't think it was called by a different name in 3.x, so it doesn't exist until 4.x? https://docs.godotengine.org/en/3.6/classes/class_projection.html

**Comment by l0g0sys** [comment]
> > No. You cannot replace any type. All types are correct in that GD script. _get_camera_projection method must return Projection.
> 
> I don't think it was called by a different name in 3.x, so it doesn't exist until 4.x? https://docs.godotengine.org/en/3.6/classes/class_projection.html

Ah, there is one but it's not public and it's called CameraMatrix in core/math source folder. So it will be more complicated to backport to 3.6.

```
bool Camera::safe_unproject_position(const Vector3 &p_pos, Point2 &r_result) const {
	ERR_FAIL_COND_V_MSG(!is_inside_tree(), false, "Camera is not inside scene.");

	Size2 viewport_size = get_viewport()->get_visible_rect().size;

	CameraMatrix cm;

	if (mode == PROJECTION_ORTHOGONAL) {
		cm.set_orthogonal(size, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
	} else {
		cm.set_perspective(fov, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
	}
	...
```

So you will need to replicate this CameraMatrix with some methods and maybe operators to make it work.
Maybe try asking in official Godot discord if anyone did that already?
You might not be only one to need this workaround for 3.x ;)

**Comment by elvisish** [comment]
> > > No. You cannot replace any type. All types are correct in that GD script. _get_camera_projection method must return Projection.
> 
> > 
> 
> > I don't think it was called by a different name in 3.x, so it doesn't exist until 4.x? https://docs.godotengine.org/en/3.6/classes/class_projection.html
> 
> 
> 
> Ah, there is one but it's not public and it's called CameraMatrix in core/math source folder. So it will be more complicated to backport to 3.6.
> 
> 
> 
> ```
> 
> bool Camera::safe_unproject_position(const Vector3 &p_pos, Point2 &r_result) const {
> 
> 	ERR_FAIL_COND_V_MSG(!is_inside_tree(), false, "Camera is not inside scene.");
> 
> 
> 
> 	Size2 viewport_size = get_viewport()->get_visible_rect().size;
> 
> 
> 
> 	CameraMatrix cm;
> 
> 
> 
> 	if (mode == PROJECTION_ORTHOGONAL) {
> 
> 		cm.set_orthogonal(size, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
> 
> 	} else {
> 
> 		cm.set_perspective(fov, viewport_size.aspect(), near, far, keep_aspect == KEEP_WIDTH);
> 
> 	}
> 
> 	...
> 
> ```
> 
> 
> 
> So you will need to replicate this CameraMatrix with some methods and maybe operators to make it work.
> 
> Maybe try asking in official Godot discord if anyone did that already?
> 
> You might not be only one to need this workaround for 3.x ;)

Thanks, I have asked a little and I've even opened a job post in the unofficial server, but I don't think anyone's interested unfortunately.

It's a shame as I make sure heavy use of the frustum camera, but I can't even really fake it either with offsets or a hidden rotation spatial because it just aims so much differently to a real rotation.

Would the entire CameraMatrix need reimplementing in gdscript?

**Comment by elvisish** [comment]
In case it's of use to anyone, I'm just using this to get a close approximation:
```swift
var center = viewport.size * 0.5
var fake_frust = viewport.size.y * 108.7866108786611 * -1
	var to_shear = cam.project_ray_origin(center) + cam.project_ray_normal(center + Vector2(0,cam.get_frustum_offset().y * fake_frust)) * ray_length
```
I don't know why `108.7866108786611` works, but it does, the magic number can probably be adjusted if it doesn't work with a particular set-up.

**Comment by bentles** [comment]
was also having issues with the frustum projection mode with an offset. The offset lets my game look a lot better but I was sad to see ray tracing was completely broken after switching... this change seems to fix that for me <3 what's the process for getting the change merged?

**Comment by Flarkk** [comment]
Rebased
